vsintegration/src/FSharp.Editor/CodeLens/FSharpCodeLensService.fs

```diff
@@ -0,0 +1,385 @@
+                                    EasingFunction = QuadraticEase()
+                                    )
+                            let sb = Storyboard()
+                            Storyboard.SetTarget(sb, uiElement)
+                            Storyboard.SetTargetProperty(sb, PropertyPath Control.OpacityProperty)
+                            sb.Children.Add animation
+                            lineLens.AddUiElementToCodeLensOnce (trackingSpan, uiElement)
+                            lineLens.RelayoutRequested.Enqueue ()
+                            sb.Begin()
+                        else
+                            logWarningf "Couldn't retrieve code lens information for %A" codeLens.FullTypeSignature
+                        // logInfo "Adding text box!"
+                    } |> StartAsyncSafe CancellationToken.None "UIElement creation"
+
+            for value in tagsToUpdate do
+                let trackingSpan, (newTrackingSpan, _, codeLens) = value.Key, value.Value
+                // logInfof "Adding ui element for %A" (codeLens.TaggedText)
+                lineLens.RemoveCodeLens trackingSpan |> ignore
+                let Grid = lineLens.AddCodeLens newTrackingSpan
+                // logInfof "Trackingspan %A is being added." trackingSpan 
+                if codeLens.Computed && (isNull codeLens.UiElement |> not) then
+                    let uiElement = codeLens.UiElement
+                    lineLens.AddUiElementToCodeLensOnce (newTrackingSpan, uiElement)
+                else
+                    Grid.IsVisibleChanged
+                    |> Event.filter (fun eventArgs -> eventArgs.NewValue :?> bool)
+                    |> Event.add (createCodeLensUIElement codeLens newTrackingSpan)
+
+            for value in codeLensToAdd do
+                let trackingSpan, codeLens = value
+                let Grid = lineLens.AddCodeLens trackingSpan
+                logInfof "Trackingspan %A is being added." trackingSpan 
+                
+                Grid.IsVisibleChanged
+                |> Event.filter (fun eventArgs -> eventArgs.NewValue :?> bool)
+                |> Event.add (createCodeLensUIElement codeLens trackingSpan)
+
+            for oldResult in oldResults do
+                let trackingSpan, _ = oldResult.Value
+                // logInfof "removing trackingSpan %A" trackingSpan
+                lineLens.RemoveCodeLens trackingSpan |> ignore
+
+            logInfof "Finished updating line lens."
+            
+            if not firstTimeChecked then
+                firstTimeChecked <- true
+        } |> Async.Ignore
+    
+    do  
+        begin
+            buffer.Changed.AddHandler(fun _ e -> (self.BufferChanged e))
+            async {
+              let mutable numberOfFails = 0
+              while not firstTimeChecked && numberOfFails < 10 do
+                  try
+                      do! executeCodeLenseAsync()
+                      do! Async.Sleep(1000)
```

Comment:
This should be a configurable constant.  And what is this anyway?  It looks like your coding an agent, but why is it needed.  A big comment is needed to justify this kind of code as it can be expensive to debug and maintain